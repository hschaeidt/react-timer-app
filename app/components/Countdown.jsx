import React from 'react';
import Clock from './Clock';
import CountdownForm from './CountdownForm';

export const STATUS = {
  STARTED: 'started',
  STOPPED: 'stopped',
};

export default class Countdown extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      count: 0,
      status: STATUS.STOPPED,
    };

    this.startTimer = this.startTimer.bind(this);
    this.handleSetCountdown = this.handleSetCountdown.bind(this);
  }

  componentDidUpdate(prevProps, prevState) {
    if (this.state.status !== prevState.status) {
      switch (this.state.status) {
        case STATUS.STARTED:
          this.startTimer();
          break;
        case STATUS.STOPPED:
          clearInterval(this.timer);
          break;
      }
    }
  }

  startTimer() {
    this.timer = setInterval(() => {
      const secondsRemaining = this.state.count - 1;
      this.setState({
        count: secondsRemaining,
      });

      if (secondsRemaining === 0) {
        this.setState({
          status: STATUS.STOPPED,
        });
      }
    }, 1000);
  }

  handleSetCountdown(seconds) {
    this.setState({
      count: seconds,
      status: STATUS.STARTED,
    });
  }

  render() {
    const { count } = this.state;

    return (
      <div>
        <Clock totalSeconds={count} />
        <CountdownForm onSetCountdown={this.handleSetCountdown} />
      </div>
    );
  }
}
