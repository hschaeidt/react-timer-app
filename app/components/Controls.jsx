import React from 'react';
import { STATUS } from './Countdown';

export default class Controls extends React.Component {
  render() {
    const { countdownStatus } = this.props;

    const renderStartStopButton = () => {
      if (countdownStatus === STATUS.STARTED) {
        return <button className="button secondary">Pause</button>;
      }

      return <button className="button primary">Start</button>;
    };

    return (
      <div className="controls">
        {renderStartStopButton()}
        <button className="button alert hollow">Clear</button>
      </div>
    );
  }
}

Controls.propTypes = {
  countdownStatus: React.PropTypes.string.isRequired,
};