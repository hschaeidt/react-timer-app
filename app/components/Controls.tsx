import * as React from 'react';
import { STATUS } from './Countdown';

interface ControlsProps extends React.Props<any> {
  countdownStatus: STATUS;
  onStatusChange: (newState: STATUS) => void;
}

export default class Controls extends React.Component<ControlsProps, void> {
  constructor(props: ControlsProps) {
    super(props);

    this.onStatusChange = this.onStatusChange.bind(this);
  }

  onStatusChange(newStatus: STATUS): () => void {
    return () => {
      this.props.onStatusChange(newStatus);
    };
  }

  render(): JSX.Element {
    const { countdownStatus } = this.props;

    const renderStartStopButton = () => {
      if (countdownStatus === STATUS.STARTED) {
        return (
          <button
            className="button secondary"
            onClick={this.onStatusChange(STATUS.PAUSED)}
          >
            Pause
          </button>
        );
      }

      return (
        <button
          className="button primary"
          onClick={this.onStatusChange(STATUS.STARTED)}
        >
          Start
        </button>
      );
    };

    return (
      <div className="controls">
        {renderStartStopButton()}
        <button
          className="button alert hollow"
          onClick={this.onStatusChange(STATUS.STOPPED)}
        >
          Clear
        </button>
      </div>
    );
  }
}
