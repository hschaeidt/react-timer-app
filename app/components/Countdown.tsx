import * as React from 'react';
import Clock from './Clock';
import CountdownForm from './CountdownForm';
import Controls from './Controls';

export enum STATUS {
  STARTED,
  STOPPED,
  PAUSED,
}

interface CountdownProps extends React.Props<any> {}

interface CountdownState {
  count: number,
  status: STATUS,
}

export default class Countdown extends React.Component<CountdownProps, CountdownState> {
  private timer: number;

  constructor(props: CountdownProps) {
    super(props);

    this.state = {
      count: 0,
      status: STATUS.STOPPED,
    };

    this.startTimer = this.startTimer.bind(this);
    this.stopTimer = this.stopTimer.bind(this);
    this.handleSetCountdown = this.handleSetCountdown.bind(this);
    this.handleStatusChange = this.handleStatusChange.bind(this);
  }

  componentDidUpdate(prevProps: CountdownProps, prevState: CountdownState): void {
    if (this.state.status !== prevState.status) {
      switch (this.state.status) {
        case STATUS.STARTED:
          this.startTimer();
          break;
        default:
          this.stopTimer();
          break;
      }
    }
  }

  componentWillUnmount(): void {
    this.stopTimer();
  }

  startTimer(): void {
    this.timer = setInterval(() => {
      const secondsRemaining = this.state.count - 1;
      this.setState({
        count: secondsRemaining,
      });

      if (secondsRemaining === 0) {
        this.handleStatusChange(STATUS.STOPPED);
      }
    }, 1000);
  }

  stopTimer(): void {
    clearInterval(this.timer);
    this.timer = undefined;
  }

  handleSetCountdown(seconds: number) {
    if (seconds > 0) {
      this.setState({
        count: seconds,
        status: STATUS.STARTED,
      });
    }
  }

  handleStatusChange(status: STATUS) {
    const state: CountdownState = { 
      status,
      count: this.state.count,
    };

    if (status === STATUS.STOPPED) {
      state.count = 0;
    }

    this.setState(state);
  }

  render() {
    const { count, status } = this.state;

    const renderControlArea = () => {
      if (status !== STATUS.STOPPED) {
        return <Controls countdownStatus={status} onStatusChange={this.handleStatusChange} />;
      }

      return <CountdownForm onSetCountdown={this.handleSetCountdown} />;
    };

    return (
      <div>
        <h1 className="page-title">Countdown App</h1>
        <Clock totalSeconds={count} />
        {renderControlArea()}
      </div>
    );
  }
}
