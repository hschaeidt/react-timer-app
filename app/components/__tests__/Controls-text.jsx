import React from 'react';
import renderer from 'react-test-renderer';
import Controls from '../Controls';
import { STATUS } from '../Countdown';

describe('<Controls />', () => {
  it('should exist', () => {
    expect(Controls).toBeDefined();
  });

  describe('render', () => {
    it('renders a pause button when started', () => {
      const tree = renderer.create(<Controls countdownStatus={STATUS.STARTED} />);
      expect(tree).toMatchSnapshot();
    });

    it('renders a start button when stopped', () => {
      const tree = renderer.create(<Controls countdownStatus={STATUS.STOPPED} />);
      expect(tree).toMatchSnapshot();
    });
  });
});
