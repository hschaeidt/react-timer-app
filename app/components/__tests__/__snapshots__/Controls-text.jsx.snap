exports[`<Controls /> render renders a pause button when started 1`] = `
<div
  className="controls">
  <button
    className="button secondary">
    Pause
  </button>
  <button
    className="button alert hollow">
    Clear
  </button>
</div>
`;

exports[`<Controls /> render renders a start button when stopped 1`] = `
<div
  className="controls">
  <button
    className="button primary">
    Start
  </button>
  <button
    className="button alert hollow">
    Clear
  </button>
</div>
`;
